### PR TITLE
Add Option<Span> to `require_lang_item`

### LINKED ISSUE #63954: Add optional span to `require_lang_item` for diagnostics with more context
[`fn require_lang_item(LangItem)`](https://github.com/rust-lang/rust/blob/615c46086a994f088c9ed569fc36df229ae115b6/src/librustc/middle/lang_items.rs#L384-L388) should also take an optional `Span` to [avoid output without context](https://github.com/rust-lang/rust/pull/63917/files#diff-94321333c9f60a3538b4f566849c232e).

### PR DESCRIPTION
**Pull request by JohnTitor**
Fixes #63954

I'm not sure where to take `Some(span)` or something so I use `None` in many places.
r? @estebank

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -1752,7 +1752,10 @@ impl<'tcx> MirPass<'tcx> for QualifyAndPromoteConstants<'tcx> {
                 fulfillment_cx.register_bound(&infcx,
                                               param_env,
                                               ty,
-                                              tcx.require_lang_item(lang_items::SyncTraitLangItem),
+                                              tcx.require_lang_item(
+                                                  lang_items::SyncTraitLangItem,
+                                                  None
```
Comment (jonas-schievink) [inline src/librustc_mir/transform/qualify_consts.rs]: This could probably point at `body.span`

File: src/librustc_mir/util/elaborate_drops.rs [hunk 2]
```diff
@@ -897,7 +897,7 @@ where
     ) -> BasicBlock {
         let tcx = self.tcx();
         let unit_temp = Place::from(self.new_temp(tcx.mk_unit()));
-        let free_func = tcx.require_lang_item(lang_items::BoxFreeFnLangItem);
+        let free_func = tcx.require_lang_item(lang_items::BoxFreeFnLangItem, None);
```
Comment (jonas-schievink) [inline src/librustc_mir/util/elaborate_drops.rs]: This could point at `self.source_info.span`

File: src/librustc_typeck/check/cast.rs [hunk 3]
```diff
@@ -649,7 +649,7 @@ impl<'a, 'tcx> CastCheck<'tcx> {
 
 impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     fn type_is_known_to_be_sized_modulo_regions(&self, ty: Ty<'tcx>, span: Span) -> bool {
-        let lang_item = self.tcx.require_lang_item(lang_items::SizedTraitLangItem);
+        let lang_item = self.tcx.require_lang_item(lang_items::SizedTraitLangItem, None);
```
Comment (jonas-schievink) [inline src/librustc_typeck/check/cast.rs]: `span` seems like the obvious choice here

File: src/librustc_typeck/check/mod.rs [hunk 4]
```diff
@@ -2622,7 +2622,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                  span: Span,
                                  code: traits::ObligationCauseCode<'tcx>)
     {
-        let lang_item = self.tcx.require_lang_item(lang_items::SizedTraitLangItem);
+        let lang_item = self.tcx.require_lang_item(lang_items::SizedTraitLangItem, None);
```
Comment (jonas-schievink) [inline src/librustc_typeck/check/mod.rs]: There's also a `span` passed as an argument here

File: src/test/ui/privacy/privacy2.stderr [hunk 5]
```diff
@@ -11,6 +11,10 @@ LL |     use bar::glob::foo;
    |                    ^^^
 
 error: requires `sized` lang_item
+  --> $DIR/privacy2.rs:14:14
+   |
+LL | pub fn foo() {}
+   |              ^
```
Comment (JohnTitor) [inline src/test/ui/privacy/privacy2.stderr]: @jonas-schievink This and privacy3 are expected?

Comment (jonas-schievink) [reply src/test/ui/privacy/privacy2.stderr]: Probably, but yeah it's not a very good error span. Feel free to remove the spans for the `Sized` lang item, the compiler needs that lang item for all sorts of checks that don't really correspond to any language features.

Comment (JohnTitor) [reply src/test/ui/privacy/privacy2.stderr]: Hmm, okay, it may cause confusing so I'd like to remove their spans.

**Pull request merged**